src/actions/BringMoveSwap.ts

```diff
@@ -0,0 +1,266 @@
+    { insideOutsideType: "inside" },
+    { insideOutsideType: "inside" },
+  ];
+
+  constructor(private graph: Graph, private type: string) {
+    this.run = this.run.bind(this);
+  }
+
+  private broadcastSource(
+    sources: TypedSelection[],
+    destinations: TypedSelection[]
+  ) {
+    if (sources.length === 1 && this.type !== "swap") {
+      // If there is only one source target, expand it to same length as
+      // destination target
+      return Array(destinations.length).fill(sources[0]);
+    }
+    return sources;
+  }
+
+  private getDecorationStyles(): DecorationTypes {
+    if (this.type === "use") {
+      return {
+        sourceStyle: this.graph.editStyles.referenced,
+        sourceLineStyle: this.graph.editStyles.referencedLine,
+        destinationStyle: this.graph.editStyles.pendingModification0,
+        destinationLineStyle: this.graph.editStyles.pendingLineModification0,
+      };
+    }
+    return {
+      sourceStyle: this.graph.editStyles.pendingModification0,
+      sourceLineStyle: this.graph.editStyles.pendingLineModification0,
+      destinationStyle: this.graph.editStyles.pendingModification1,
+      destinationLineStyle: this.graph.editStyles.pendingLineModification1,
+    };
+  }
+
+  private async decorateTargets(
+    sources: TypedSelection[],
+    destinations: TypedSelection[]
+  ) {
+    const decorationTypes = this.getDecorationStyles();
+    await Promise.all([
+      displayPendingEditDecorations(
+        sources,
+        decorationTypes.sourceStyle,
+        decorationTypes.sourceLineStyle
+      ),
+      displayPendingEditDecorations(
+        destinations,
+        decorationTypes.destinationStyle,
+        decorationTypes.destinationLineStyle
+      ),
+    ]);
+  }
+
+  private getEdits(
+    sources: TypedSelection[],
+    destinations: TypedSelection[]
+  ): Edit[] {
+    const usedSources: TypedSelection[] = [];
+    return flatten(
+      zip(sources, destinations).map(([source, destination]) => {
+        if (source == null || destination == null) {
+          throw new Error("Targets must have same number of args");
+        }
+
+        // Add destination edit
+        const result = [
+          {
+            editor: destination.selection.editor,
+            range: destination.selection.selection,
+            newText: source.selection.editor.document.getText(
+              source.selection.selection
+            ),
+            targetsIndex: 0,
+            originalSelection: destination,
+          },
+        ];
+
+        // Add source edit for move and swap
+        // Prevent multiple instances of the same expanded source.
+        if (this.type !== "use" && !usedSources.includes(source)) {
+          let newText = ""; // Defaults to delete source/move
+          if (this.type === "swap") {
+            newText = destination.selection.editor.document.getText(
+              destination.selection.selection
+            );
+          }
```

Comment:
hmm this is one place where the branches I think make things slightly more confusing, but I can definitely follow the code and the comments help a ton so it's probably fine

Reply:
Yeah agree but I think its fine.